# Linear Programming Example using Python and Scipy

Linear programming is a mathematical technique used to optimize a linear objective function subject to linear constraints. There are many real-world situations where linear programming can be applied, such as production planning, transportation, and resource allocation.

In this Jupyter Notebook, I showcase an example of situation where linear programming can be applied and how we can use the Python library Scipy to optimize the results.

## Transportation Planning

A use case of linear programming is in transportation planning. Let's say we have several warehouses that supply goods to multiple customers. We need to decide how much of each product to transport from each warehouse to each customer, subject to constraints such as transportation capacity and demand. We can use Scipy's `linprog` function to optimize this problem and find the optimal transportation plan.

## Distribution of Air-Conditioning Units

Let's consider a North American air-conditioning firm that has production plants in Portland and Flint. The firm needs to supply a certain number of units to its distribution centers located in Los Angeles and Atlanta. The following table shows the delivery costs, supply, and demand for each location:

| Location         | Delivery Cost ($/unit) | Supply (units) | Demand (units) |
|------------------|-----------------------|----------------|----------------|
| Portland         | 2                     | 150            |                |
| Flint            | 4                     | 200            |                |
| Los Angeles      | 3                     |                |              100 |
| Atlanta          | 5                     |                |              150 |

Our goal is to minimize the total delivery cost by determining how many units should be delivered from each production plant to each distribution center. We can model this as an integer linear programming problem, where the decision variables are the number of units to be shipped from each plant to each distribution center.

Let $X_{ij}$ be the number of units shipped from plant i (Portland or Flint) to distribution center j (Los Angeles or Atlanta). Then, we want to minimize the total delivery cost, which can be expressed as:



### Decision variables:
$X_{11}$ = Units delivered from Portland to Los Angeles.\
$X_{21}$ = Units delivered from Flint to Los Angeles.\
$X_{12}$ = Units delivered from Portland to Atlanta.\
$X_{22}$ = Units delivered from Flint to Atlanta.\

### Objective Function:

Minimize: $$X_{11} + X_{21} + X_{12} + X_{22}$$

### Constraints:

subject to the following constraints:

- Supply constraints: the total number of units shipped from each plant cannot exceed its supply:

$$X_{11} + X_{21} <= 150$$\
$$X_{12} + X_{22} <= 200$$

- Demand constraints: the total number of units shipped to each distribution center must meet its demand:

$$X_{11} + X_{21} >= 100$$\
$$X_{12} + X_{22} >= 150$$

- Non-negativity constraints: the number of units shipped cannot be negative:

$$X_{11},X_{21},X_{12},X_{22} >= 0$$

In [2]:
'''We can use Scipy's `linprog` function to 
solve this problem. First, we need to transform 
the problem into standard form by converting the
objective function to a maximization problem 
and adding slack variables to the constraints.
Here's the complete code:
'''

from scipy.optimize import linprog

# Objective coefficients
c = [1, 1, 1, 1]

# Left-hand side matrix of constraints
A = [[1, 0, 1, 0],
     [0, 1, 0, 1],
     [-1, 0, 0, 0],
     [0, -1, 0, 0],
     [0, 0, -1, 0],
     [0, 0, 0, -1]]

# Right-hand side vector of constraints
b = [150, 200, -100, -150, 0, 0]

# Bounds on decision variables
bounds = [(0, None), (0, None), (0, None), (0, None)]

# Solve the linear programming problem
res = linprog(c=c, A_ub=A, b_ub=b, bounds=bounds, method='simplex')

# Print the optimal solution
print(res)



     con: array([], dtype=float64)
     fun: 250.0
 message: 'Optimization terminated successfully.'
     nit: 4
   slack: array([50., 50.,  0.,  0.,  0.,  0.])
  status: 0
 success: True
       x: array([100., 150.,   0.,   0.])


### Conclusion & Results

The output shows the optimal solution, which tells us how many units should be shipped from each plant to each distribution center:

$$X_{11} = 100 $$\
$$X_{21} = 150 $$\
$$X_{12} = 0 $$\
$$X_{22} = 0$$\
$$Z = 250 $$


100 units from Portland to Los Angeles, \
150 units from Flint to Los Angeles, \
Shipping to Atlanta would not be cost efficent if you are looking to minimize cost. \
Total cost of $250.